### ISSUE TITLE
Registering transaction on newrelic

### ISSUE THREAD

**Issue by rt3norio** [rebing/graphql-laravel#670]
### Versions:
- graphql-laravel Version: ^4.0
- Laravel Version: v5.8.37
- PHP Version: 7.1

### Question:

Hello!
We are implementing observability through NewRelic APM features...

But after installing on project, the NewRelic reported the following as the only trace name: 
**/graphql.gateway.post**

Turns out it registers all the transactions as the same.

To solve this problem (that we already had in other projects with Yii) we implemented in the superclass **MutationAbstract** an BeforeResolve, that grabs the information that we need, such as:
- Query/Mutation name
- Query/Mutation params
- user
- etc
sets the transaction name, params, and so on...

Here, in this Laravel project, **we didn't create such class**, and all queries and mutations extends directly from **rebing** classes.

We thought a better approach should be using a middleware to grab such information, but after some testing I realized that the middleware doesn't have access to if it was a query or mutation, the params, and so on...

after looking in https://github.com/rebing/graphql-laravel/issues/66 with no luck, I would like some suggestion to how to implement such observability

Thanks!

**Comment by mfn** [comment]
Very good question. I don't have an immediate answer for you.

There's also work on https://github.com/rebing/graphql-laravel/pull/594 , maybe the is suitable but I can't say.

Just to get a better picture: do you have an example how this looks before/after? I'm using sentry and I also see it mostly reports something like this:
![image](https://user-images.githubusercontent.com/87493/91754622-06e7d800-ebca-11ea-8161-2867df495c71.png)

But so far I didn't bother because it also includes this, which so far was sufficient:
![image](https://user-images.githubusercontent.com/87493/91754667-1830e480-ebca-11ea-81ab-9ff3d64425b0.png)

**Comment by rt3norio** [comment]
Giving you a better picture, on the newrelic dashboard, we have a view for the 'transactions', and they are grouped by transaction name.

The transaction name, by default, is the URI of the request.

and on my dashboard it looks like this:
![ksnip_20200901-120302](https://user-images.githubusercontent.com/9087474/91872750-98992900-ec4e-11ea-9ac3-3b86358a1d04.png)

to solve this problem, the documentation says that we should use this php function to register the name of the transaction, together with newrelic_add_custom_parameter to do what is obvious.
https://docs.newrelic.com/docs/agents/php-agent/php-agent-api/newrelic_name_transaction 

and it splits into corresponding transactions
![ksnip_20200901-120528](https://user-images.githubusercontent.com/9087474/91873466-6cca7300-ec4f-11ea-81fb-2c79ed1b9d41.png)

and this names was manually added on the abstract class like this:
> newrelic_name_transaction ('GraphQL' . '/' .'Mutation' . '/' . $this->name());

And, newrelic doesn't save the body data, and even if it does, it messes with the metrics provided. 
the best approach is to name correctly the transaction.

**Comment by rt3norio** [comment]
And, just to leave an update of what I'm having to do to circunvent this middleware issue, this is the code to intercept the resolver.



    abstract class AbstractQuery extends Query
    {
        abstract public function resolver($root, $args);
        
        protected function beforeResolve($root, $args)
        {
            $this->setNewrelicTransaction();
        }

        final protected function resolve($root, $args)
        {
            $this->beforeResolve($root, $args);
            return $this->resolver($root, $args);
        }
    }

I hope #594 get merged soon so I can use middlewares to do it.
Thanks @mfn for the quick response!

**Comment by mfn** [comment]
Curious question: how would you deal with batch requests?

Aka a single `POST` request including multiple queries/mutations?

**Comment by rt3norio** [comment]
We don't use such things on our business...
but...
@alanwillms @renatofl any ideas?
Summoning the big guys to help me out...

**Comment by mfn** [comment]
I'm quoting this from the original issue description.
> We thought a better approach should be using a middleware to grab such information

To the best of my knowledge, this was referring to "Laravel middlewares" which indeed won't have access to this details (at least not easily).

In the meantime, also what I mentioned in https://github.com/rebing/graphql-laravel/issues/670#issuecomment-683957385 , the support for "GraphQL middleware" has been merged.

Aka it has the same signature as `resolve` and can access all the information there, including the query/mutation name. In fact I've already used this as advantage in a (private) project to things like:
- additional security protection for specific ("approve list") queries and mutations
- wrap mutations in a custom application specific transaction flow

Given that flexibility, I believe writing a middleware which sets your desired transactions for newrelic (or sentry etc.) should be the right way to go.

In this spirit, I'm closing this issue as I believe an appropriate solution exists now.

Thank you!
